datasource/src/main/scala/quasar/physical/s3/S3DatasourceModule.scala

```diff
@@ -41,16 +43,19 @@ import slamdata.Predef.{Stream => _, _}
 object S3DatasourceModule extends LightweightDatasourceModule {
   def kind: DatasourceType = s3.datasourceKind
 
-  def lightweightDatasource[F[_]: ConcurrentEffect: MonadResourceErr: Timer](config: Json)
-      : F[InitializationError[Json] \/ Disposable[F, Datasource[F, Stream[F, ?], ResourcePath]]] = {
+  def lightweightDatasource[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr: Timer](
+    config: Json)(implicit ec: ExecutionContext)
+      : F[InitializationError[Json] \/ Disposable[F, Datasource[F, Stream[F, ?], ResourcePath]]] =
     config.as[S3Config].result match {
-      case Right(s3Config) => {
-        Http1Client[F]() flatMap { client =>
+      case Right(s3Config) =>
+        val clientResource = BlazeClientBuilder[F](ec).resource
+
+        clientResource.use { client =>
           val s3Ds = new S3Datasource[F](client, s3Config)
           val ds: Datasource[F, Stream[F, ?], ResourcePath] = s3Ds
 
           s3Ds.isLive.ifM({
-            Disposable(ds, client.shutdown).right.pure[F]
+            Disposable(ds, s3.resourceCleanup(clientResource)).right.pure[F]
```

Comment:
Not sure if I should extract the cleanup from the `Resource` or simply return `().point[F]`. Clean up should happen regardless, since this is all wrapped in `use` anyway

Reply:
Yeah I *think* this is fine as-is.

Reply:
We definitely need the real finalizer in the returned `Disposable` as the scope of the resource is user-defined, i.e. we want to release if the server stops _or_ if the datasource is ever unmounted, the latter of which can happen at any time.

I'm actually not sure how `use` will affect this, my concern is it would free the resource too early, but I'm not convinced of it yet. I wonder if it'd be better to use the `stream` constructor with an interrupt and then compile it into an effect.

Reply:
@jsantos17 yes it is, also in the repl via `ds remove <uuid>`